Step 4

ilbender opened the pull request.

ilbender committed:
tests

ilbender committed:
.

bmax-ltx commented on app/src/androidTest/java/com/lightricks/feedexercise/data/callFeedRepoTests.kt:
```diff
@@ -2,37 +2,32 @@ package com.lightricks.feedexercise.data
 import org.junit.Before
 import org.junit.Test
 
 class callFeedRepoTests {
```
```suggestion
class CallFeedRepoTests {
```

bmax-ltx replied:
Why are there both `FeedRepositoryTest` and `CallFeedRepoTests`?

ilbender replied:
so that i can pass params to the test to make it more generic(such as different mock FeedApiServices), thats how I did it in the first exercise, is there a better way to do it?

bmax-ltx commented on app/src/androidTest/java/com/lightricks/feedexercise/data/ViewModelTest.kt:
```diff
@@ -9,38 +9,86 @@ import com.lightricks.feedexercise.database.UserProject
-class ViewModelTest (private var db : FeedDatabase,
-                     private var feedRepository: FeedRepository,
-                     private var userProjectList : List<UserProject>){
+class ViewModelTest(
```
Why are there both `ViewModelTest` and `CallViewModelTests`?

ilbender replied:
so that i can pass params to the test to make it more generic(such as different mock FeedApiServices), thats how I did it in the first exercise, is there a better way to do it?

bmax-ltx commented on app/src/androidTest/java/com/lightricks/feedexercise/data/Constans.kt:
```diff
@@ -0,0 +1,29 @@
+
+import com.lightricks.feedexercise.database.UserProject
+
+object Constans {
```
```suggestion
object Constants {
```

bmax-ltx replied:
Typo here.

bmax-ltx reviewed (commented):
A couple of questions before I review it.

bmax-ltx commented on app/src/androidTest/java/com/lightricks/feedexercise/data/MockApis.kt:
```diff
@@ -0,0 +1,46 @@
+
+
+object ConstantItemsApiService : FeedApiService {
+    private val fullJson: TemplatesMetadata = JsonToTemplateMetaData.getFeedData()
```
Let's rename the property since `getFeedData()` returns `TemplatesMetadata` not a JSON.

ilbender committed:
merged the call and test classes

bmax-ltx commented on app/src/androidTest/java/com/lightricks/feedexercise/data/ViewModelTest.kt:
```diff
@@ -1,46 +1,100 @@
+
+    @Test
+    fun testViewModelInitData() {
+        init()
```
You don't need to call `init()` in every test.
The `@Before` annotation does it for you.

bmax-ltx commented on app/src/androidTest/java/com/lightricks/feedexercise/data/ViewModelTest.kt:
```diff
@@ -1,46 +1,100 @@
+        init()
+        val viewModelFactory = FeedViewModelFactory(constantfeedRepository)
+        viewModel = viewModelFactory.create(FeedViewModel::class.java)
+        Thread.sleep(100)
```
Why do we need to sleep?

bmax-ltx commented on app/src/androidTest/java/com/lightricks/feedexercise/data/ViewModelTest.kt:
```diff
@@ -1,46 +1,100 @@
+        viewModel = viewModelFactory.create(FeedViewModel::class.java)
+        Thread.sleep(100)
+        lateinit var viewModelLst: List<FeedItem>
+        viewModel.getFeedItems().observe(
```
There is a simpler way to get the value of a LiveData in tests:
```
private fun <T> LiveData<T>.blockingObserve(): T? {
    var value: T? = null
    val latch = CountDownLatch(1)
    val observer = object : Observer<T> {
        override fun onChanged(t: T) {
            value = t
            latch.countDown()
            removeObserver(this)
        }
    }

    observeForever(observer)
    latch.await(5, TimeUnit.SECONDS)
    return value
}
```

bmax-ltx commented on app/src/androidTest/java/com/lightricks/feedexercise/data/ViewModelTest.kt:
```diff
@@ -1,46 +1,100 @@
+        val viewModelFactory = FeedViewModelFactory(constantfeedRepository)
+        viewModel = viewModelFactory.create(FeedViewModel::class.java)
+        Thread.sleep(100)
+        deleteFirstProject()
```
If this test wants to check that the VM reacts to a change in the data, then it should probably get feed items twice: before deleting and after.

bmax-ltx reviewed (changes requested):
A couple of comments.

ilbender committed:
changed the delete test in ViewModel to verify the data before the delete.

no longer uses Thread.sleep(100) to wait for IO operations so that the tests wont be flaky

renamed getFeedData() in JsonToTemplateMetaData

bmax-ltx commented on app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt:
```diff
@@ -8,34 +8,46 @@ import com.lightricks.feedexercise.database.UserProject
+
+    @Test
+    fun testGetAll(){
+        feedRepository.refresh().test().awaitTerminalEvent()
```
This test is not supposed to call `refresh()`.
1. Insert the items directly into the DB
2. Call `feedRepository.getAllProjects()`
3. Assert you got the expected items

bmax-ltx commented on app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt:
```diff
@@ -8,34 +8,46 @@ import com.lightricks.feedexercise.database.UserProject
-    fun testRefreshSaves(){
+    fun testRefreshSaves() {
         feedRepository.refresh().test().awaitTerminalEvent()
-        var fetchedList : List<UserProject>
```
Let's verify that before calling `feedRepository.refresh()` there were no items in the DB.

bmax-ltx reviewed (changes requested):
Let's fix FeedRepositoryTest.

ilbender committed:
1. changed TestRefreshSaves so that we verify the repo is empty before the refresh

2. changed getAll tests so that we insert the data directly into the db rather than calling refresh

bmax-ltx commented on app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt:
```diff
@@ -8,34 +8,49 @@ import com.lightricks.feedexercise.database.UserProject
+        feedRepository.getAllProjects().doOnNext { userProjects ->
+            fetchedList = userProjects
+            Assert.assertTrue(fetchedList.isEmpty())
+        }.subscribe()
```
This doesn't wait until userProjects are loaded.
But you can use test observer for that (like with `refresh()`).

bmax-ltx commented on app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt:
```diff
@@ -8,34 +8,49 @@ import com.lightricks.feedexercise.database.UserProject
+        feedRepository.getAllProjects().subscribe { fetchedUserProjects ->
+            fetchedList = fetchedUserProjects
+            Assert.assertEquals(fetchedList, userProjectList)
+        }
```
Also, I'm afraid the test can finish before `fetchedUserProjects` are fetched.
You should call `test().awaitTerminalEvent()`.

ilbender committed:
changed as per baruch's instructions so that asserts will happen after all fetches are expected to end.

bmax-ltx reviewed (approved):
Kudos!

bmax-ltx commented:
Just fix the typo in `Constants` please.
